### PR TITLE
feat(notifications): turn plugins into dynamic ones

### PR DESCRIPTION
**Pull request by mareklibra**
Fixes [FLPATH-877](https://issues.redhat.com//browse/FLPATH-877)

### CODE REVIEW

File: plugins/notifications-backend/src/service/router.ts [hunk 1]
```diff
@@ -23,7 +25,14 @@ import { RouterOptions } from './types';
 export async function createRouter(
   options: RouterOptions,
 ): Promise<express.Router> {
-  const { logger, dbConfig, catalogClient } = options;
+  const { logger, database, discovery, config } = options;
+
+  // workaround for creating the database when client is not sqlite
```
Comment (gashcrumb) [inline plugins/notifications-backend/src/service/router.ts:30]: Is this a normal thing?  Is there an issue for this somewhere?

Comment (mareklibra) [reply plugins/notifications-backend/src/service/router.ts:30]: Nope.  I fixed that in the upstream PR:
https://github.com/backstage/backstage/pull/22099/files#diff-72c8723c8f4d71d407e124d229ed18258167dc4547c9babc7ac3faac3226b413R36

and plan to cherry-pick the change to this PR as well.
But in an independent PR.

File: plugins/notifications-backend/src/service/router.ts [hunk 2]
```diff
@@ -23,7 +25,14 @@ import { RouterOptions } from './types';
...
+  const dbClientWorkaround = await database.getClient();
```
Comment (ydayagi) [inline plugins/notifications-backend/src/service/router.ts:31]: please don't call it workaround. it only makes people notice it. it is what it is. a db client

File: plugins/notifications-backend/config.d.ts
Comment (ydayagi) [inline plugins/notifications-backend/config.d.ts:1]: what is the reason for this interface? and why not simply put it in types.ts?

File: plugins/notifications-backend/README.md [hunk 3]
```diff
@@ -80,25 +92,14 @@ import { PluginEnvironment } from '../types';
...
     identity: env.identity,
     logger: env.logger,
     permissions: env.permissions,
     tokenManager: env.tokenManager,
-    dbConfig,
-    catalogClient,
-    externalCallerSecret,
+    database: env.database,
+    discovery: env.discovery,
+    config: env.config,
```
Comment (ydayagi) [inline plugins/notifications-backend/README.md:102]: what is the reason for adding the whole config?

File: plugins/notifications-backend/README.md [hunk 4]
```diff
@@ -80,25 +92,14 @@ import { PluginEnvironment } from '../types';
 export default async function createPlugin(
   env: PluginEnvironment,
 ): Promise<Router> {
-  // workaround for creating the database when client is not sqlite
-  const dbClient = await env.database.getClient()
-  dbClient.destroy()
-
-  const catalogClient = new CatalogClient({ discoveryApi: env.discovery });
```
Comment (ydayagi) [inline plugins/notifications-backend/README.md:87]: why replace this line?

File: plugins/notifications-backend/README.md [hunk 5]
```diff
@@ -80,25 +92,14 @@ import { PluginEnvironment } from '../types';
...
-
   return await createRouter({
     identity: env.identity,
     logger: env.logger,
     permissions: env.permissions,
     tokenManager: env.tokenManager,
-    dbConfig,
-    catalogClient,
-    externalCallerSecret,
+    database: env.database,
```
Comment (ydayagi) [inline plugins/notifications-backend/README.md:100]: what is the reason for passing the database?

File: plugins/notifications-backend/src/service/router.ts (hunk 2 above)
Comment (ydayagi) [inline plugins/notifications-backend/src/service/router.ts:31]: this should not be here. the db creation whether required or not (sqlite does not reuqire it) is not part of the plugin. is it a limitation due to dynamic plugin?

File: plugins/notifications-backend/README.md (hunk 5 above)
Comment (mareklibra) [reply plugins/notifications-backend/README.md:100]: The dynamic plugins call the `createRouter` and provide it with the references. We can not customize it as we did in our implementatino of the `createPlugin`

File: plugins/notifications-backend/src/service/router.ts (hunk 2 above)
Comment (mareklibra) [reply plugins/notifications-backend/src/service/router.ts:31]: yes, to unify the flows

Comment (mareklibra) [reply plugins/notifications-backend/src/service/router.ts:31]: We have actual `dbClient` below. So renaming

**Pull request merged**